google/cloud/spanner_dbapi/parse_utils.py

```diff
@@ -264,9 +256,7 @@ def parse_insert(insert_sql, params):
     :param params: A list of parameters.
 
     :rtype: dict
-    :returns: A dictionary that maps `sql_params_list` to the list of
-              parameters in cases a), b), d) or the dictionary with information
-              about the resulting table in case c).
+    :returns: A list of (sql, param) tuples
     """  # noqa
     match = RE_INSERT.search(insert_sql)
```

Comment:
I know that this is not part of this change, but I think this part should be more lenient. As far as I can tell, the function will return an error if it cannot match the insert regexp, and then refuse to execute the statement(s). I would rather expect something like:
1. Try to parse the insert statement, and if it does not work, fall back to a 'simple' case where the sql statement and parameter list is converted to a normal batch update RPC invocation. That prevents failures when an application tries to execute an insert statement that does not match the regexp (for example if it contains comments, hints, ...)
2. If the parse succeeds, then try the different scenarios in this function.
3. If the parse fails because it encounters some other non-expected scenario, then also fall back to the default of just trying to execute the statements as a batch update.

The above would ensure that the common cases of (generated) sql are parsed into a more efficient format, while other sql is still executed without any unexpected errors.

It might also be good to consider making entire parsing optional. It might be surprising for users that dbapi actually changes the SQL that is executed. Normally a driver like dbapi should execute the SQL as is. The generation of efficient SQL statements should as far as possible be a responsibility of any framework or application that uses the driver.

Reply:
After looking more closely, this parsing and rewriting of insert statements doesn't seem to add any benefits. It's only used for `execute` which means at best it was returning the single SQL statement and at worst it was splitting it into multiple statements.

I have removed `parse_insert` in favor of sending the SQL as is.